File: src/main/java/io/github/tamawish/pureeconomy/listener/PlayerConnectionListener.java
package io.github.tamawish.pureeconomy.listener;

import io.github.tamawish.pureeconomy.PureEconomy;
import io.github.tamawish.pureeconomy.util.UpdateChecker;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;

/** Creates accounts on join and flushes balances on quit. */
public final class PlayerConnectionListener implements Listener {

  private final PureEconomy plugin;
  private final UpdateChecker updateChecker;

  /**
   * Creates a connection listener.
   *
   * @param plugin owning plugin instance
   * @param updateChecker optional release notifier; may be null when disabled later
   */
  public PlayerConnectionListener(PureEconomy plugin, UpdateChecker updateChecker) {
    this.plugin = plugin;
    this.updateChecker = updateChecker;
  }

  /**
   * Ensures a wallet exists for the joining player and shows update notices to admins.
   *
   * @param event join event from the server
   */
  @EventHandler
  public void onJoin(PlayerJoinEvent event) {
    Player player = event.getPlayer();
    if (plugin.getConfig().getBoolean("create-on-join", true)) {
      plugin.economy().ensureAccount(player.getUniqueId(), player.getName());
    }
    if (updateChecker != null) {
      updateChecker.notifyPlayerIfNeeded(player);
    }
  }

  /**
   * Persists the quitting player's account asynchronously.
   *
   * @param event quit event from the server
   */
  @EventHandler
  public void onQuit(PlayerQuitEvent event) {
    plugin.economy().savePlayerAsync(event.getPlayer().getUniqueId());
  }
}

File: src/main/kotlin/io/github/tamawish/pureeconomy/listener/PlayerConnectionListener.kt
package io.github.tamawish.pureeconomy.listener

import io.github.tamawish.pureeconomy.PureEconomy
import io.github.tamawish.pureeconomy.util.UpdateChecker
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent

/** Creates accounts on join and flushes balances on quit. */
class PlayerConnectionListener(
    private val plugin: PureEconomy,
    private val updateChecker: UpdateChecker?,
) : Listener {
    /** Ensures a wallet exists for the joining player and shows update notices to admins. */
    @EventHandler
    fun onJoin(event: PlayerJoinEvent) {
        val player = event.player
        if (plugin.settings().getBoolean("create-on-join", true)) {
            plugin.economy().ensureAccountAsync(player.uniqueId, player.name)
        }
        updateChecker?.notifyPlayerIfNeeded(player)
    }

    /** Persists the quitting player's account asynchronously. */
    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        plugin.economy().unloadPlayer(event.player.uniqueId)
    }
}
